### ISSUE TITLE
`bfToBits` and `repIsSubnormal` loop when called on 0 on M2 Mac

### ISSUE THREAD

**Issue by danmatichuk** [GaloisInc/libBF-hs#22]
Environment: M2 Mac running OS X 13.4.1
ghc: 9.2.7

Steps to reproduce:
```
ghci> :set -package libBF
ghci> import LibBF
ghci> bfIsSubNormal (float32 NearEven) bfPosZero
^CInterrupted.
ghci> bfIsSubnormal (float32 NearEven) (bfFromInt 1)
False
ghci> bfIsSubnormal (float32 NearEven) (bfFromInt 0)
^CInterrupted.
ghci> bfIsSubnormal (float32 NearEven) bfNaN
False
ghci> bfIsSubnormal (float32 NearEven) bfNegInf
False
ghci> bfIsSubnormal (float32 NearEven) bfPosInf
False
ghci> 
```

**Comment by RyanGlScott** [comment]
Wow, this is a juicy one. My first inclination was to suspect `libBF`'s use of `foreign import`s, since dodgy `foreign import`s have caused bugs on AArch64 Macs in the past (see [this blog post](https://www.haskell.org/ghc/blog/20210709-capi-usage.html)). But looking at [`libbf.h`](https://github.com/GaloisInc/libBF-hs/blob/98d71430185d708accd9e23969b66c8a190a783d/libbf-2020-01-19/libbf.h), there are no variadic functions (one of the things called out in that blog post), and the `foreign import`s for the functions that `bfIsSubnormal`, `bfPosZero`, and `bfFromInt` use appear to be correct upon a quick glance. As such, I don't have any solid leads at the moment.

At this point, I think it would be helpful to get a better sense of where the loop happens. Since you are able to Ctrl-C it, my hunch is that the loop occurs in Haskell code, not in C code, as the latter is harder to interrupt under normal circumstances.

**Comment by RyanGlScott** [comment]
FWIW, I asked about this on the #ghc IRC channel, and someone confirmed that this issue also affects AArch64 Linux, not just macOS.

**Comment by danmatichuk** [comment]
Here is a minimal fix at least:
```
bfToBits :: BFOpts -> BigFloat -> Integer
bfToBits opts bf 
  | bfIsZero bf = 0
  | otherwise = res
  where
  ...
  ```

**Comment by danmatichuk** [comment]
Actually technically it should be something like this:
```
bfToBits opts bf 
  | bfIsZero bf, bfIsPos bf = 0
  | bfIsZero bf, bfIsNeg bf, 15 <- (e + p') = (fromIntegral (maxBound :: Int16)) + 1
  | bfIsZero bf, bfIsNeg bf, 31 <- (e + p') = (fromIntegral (maxBound :: Int32)) + 1
  | bfIsZero bf, bfIsNeg bf, 63 <- (e + p') = (fromIntegral (maxBound :: Int64)) + 1
  | otherwise = res
  ```

**Comment by RyanGlScott** [comment]
I figured out what is causing this. It's this code:

https://github.com/GaloisInc/libBF-hs/blob/98d71430185d708accd9e23969b66c8a190a783d/libbf-2020-01-19/libbf.h#L30-L34

Guarding this definition with `defined(__x86_64__)` is quite sketchy, as nothing about this code is specific to x86-64. Really, the first definition should be used if we are on any 64-bit architecture, such as AArch64. (I gather that `libbf` was developed and tested on a x86-64 machine, but still.)

Clearly we should change this to accommodate architectures like AArch64, but _how_ should we change it? We are using the latest version of `libbf` from the website [here](https://bellard.org/libbf/), but despite this, the same author is maintaining a newer version of `libbf` as part of the `quickjs` project [here](https://github.com/bellard/quickjs). `quickjs`'s version of `libbf` checks for 64 bits in a [smarter way](https://github.com/bellard/quickjs/blob/2788d71e823b522b178db3b3660ce93689534e6d/libbf.h#L30-L34):

```c
#if INTPTR_MAX >= INT64_MAX
#define LIMB_LOG2_BITS 6
#else
#define LIMB_LOG2_BITS 5
#endif
```

I managed to reproduce the bug in an AArch Linux Docker container, and after replacing the vendored version of `libbf` in this library with the one in the `quickjs` repo, I've confirmed that the bug goes away. Therefore, I propose we fix this bug by updating `libbf`.

**Comment by RyanGlScott** [comment]
I've uploaded [`libBF-0.6.6`](https://hackage.haskell.org/package/libBF-0.6.6) to Hackage, which includes a fix for this issue (#24).
